handler/src/main/java/io/netty/handler/ssl/ApplicationProtocolNegotiationHandler.java

```diff
@@ -91,20 +93,31 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
                     String protocol = sslHandler.applicationProtocol();
                     configurePipeline(ctx, protocol != null ? protocol : fallbackProtocol);
                 } else {
-                    handshakeFailure(ctx, handshakeEvent.cause());
+                    // if the event is not produced because of an successful handshake we will receive the same
+                    // exception in exceptionCaught(...) and handle it there. This will allow us more fine-grained
+                    // control over which exception we propagate down the ChannelPipeline.
+                    //
+                    // See https://github.com/netty/netty/issues/10342
                 }
             } catch (Throwable cause) {
                 exceptionCaught(ctx, cause);
             } finally {
-                ChannelPipeline pipeline = ctx.pipeline();
-                if (pipeline.context(this) != null) {
-                    pipeline.remove(this);
+                // Handshake failures are handled in exceptionCaught(...).
+                if (handshakeEvent.isSuccess()) {
```

Comment:
@normanmaurer tbh I would remove the isSuccess check, I think it would be safer to always attempt to remove the handler from the pipeline, what if an exception is not thrown but handshake event is false?

Reply:
thats impossible I think